File: 704-binary-search/704-binary-search.py


class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        
        if nums is None or len(nums) == 0:
            return -1;
        
        if len(nums) == 1:
            if nums[0] == target:
                return 0
            else:
                return -1
        
        return self.searchNums(nums, 0, len(nums)-1, target)
    
    def searchNums(self, nums, begin, end, target):
        
        """
        :type nums: List[int]
        :type target: int
        :type begin: int index
        :type end: int index
        :rtype: int
        """
        
        if nums is None or end-begin == -1:
            return -1;
        
        if end-begin == 0:
            if nums[begin] == target:
                return begin
            else:
                return -1
        
        if end-begin == 1:
            if nums[begin] == target:
                return begin
            elif nums[end] == target:
                return end
            else:
                return -1
                
        if nums[(end-begin)/2+begin] == target:
            return (end-begin)/2+begin
        elif nums[(end-begin)/2+begin] > target:
            return self.searchNums(nums, begin, (end-begin)/2+begin-1, target)
        else:
            return self.searchNums(nums, (end-begin)/2+begin, end, target)

File: 704-binary-search/704-binary-search.java

class Solution {
    public int search(int[] nums, int target) {
        int[] index = new int[nums.length];
        for (int i=0; i<nums.length; i++){
            index[i] = i;
        }

        return search(nums,index,target);
        
    }
    
    public int search(int[] nums, int[] index, int target){
    
                
        int index1 = 0;
        int index2 = 0;
        
        if (nums.length == 1){
            if (nums[0]==target) return index[0];
            else return -1;
        }
        else if (nums.length == 2){
            if (nums[0]==target) return index[0];
            else if (nums[1]==target) return index[1];
            else return -1;
        }

        else if (nums.length%2 == 0){
            index1 = search(subArray(nums,0,nums.length/2),
                   subArray(index,0,index.length/2),target);
            index2 = search(subArray(nums,nums.length/2,nums.length),
                   subArray(index,index.length/2,index.length),target);
        }
        else {
            index1 = search(Arrays.copyOfRange(nums,0,nums.length/2),
                   Arrays.copyOfRange(index,0,index.length/2),target);
            index2 = search(Arrays.copyOfRange(nums,nums.length/2,nums.length),
                   Arrays.copyOfRange(index,index.length/2,index.length),target);
        }
        
        if (index1 > -1) return index1;
        else return index2;
    }
    
    public int[] subArray (int[] arr, int start, int end){
        int[] newArr = new int [end-start];
        for (int i=0; i<newArr.length;i++){
            newArr[i] = arr[start];
            start++;
        }
        return newArr;
    }

}